src/malli/experimental.cljc

```diff
@@ -1,60 +1,94 @@
-        ->schema (fn [{:keys [schema]}] [:=> schema (:schema return :any)])
-        single (= :single (key arities))
-        parglists (if single (->> arities val parse vector) (->> arities val :arities (map parse)))
-        raw-arglists (map :raw-arglist parglists)
-        schema (as-> (map ->schema parglists) $ (if single (first $) (into [:function] $)))]
-    `(let [defn# (c/defn
-                   ~name
-                   ~@(some-> doc vector)
-                   ~(assoc meta :raw-arglists (list 'quote raw-arglists), :schema schema)
-                   ~@(map (fn [{:keys [arglist prepost body]}] `(~arglist ~prepost ~@body)) parglists)
-                   ~@(when-not single (some->> arities val :meta vector)))]
-       (m/=> ~name ~schema)
-       defn#)))
+#?(:clj
+   (c/defn -defn [schema args]
+     (let [{:keys [name return doc meta arities] :as parsed} (m/parse schema args)
+           _            (when (= ::m/invalid parsed) (m/-fail! ::parse-error {:schema schema, :args args}))
+           parse        (fn [{:keys [args] :as parsed}] (merge (md/parse args) parsed))
+           ->schema     (fn [{:keys [schema]}] [:=> schema (:schema return :any)])
+           single       (= :single (key arities))
+           parglists    (if single (->> arities val parse vector) (->> arities val :arities (map parse)))
+           raw-arglists (map :raw-arglist parglists)
+           schema       (as-> (map ->schema parglists) $ (if single (first $) (into [:function] $)))]
+       `(let [defn# (c/defn
+                      ~name
+                      ~@(some-> doc vector)
+                      ~(assoc meta :raw-arglists (list 'quote raw-arglists), :schema schema)
+                      ~@(map (fn [{:keys [arglist prepost body]}] `(~arglist ~prepost ~@body)) parglists)
+                      ~@(when-not single (some->> arities val :meta vector)))]
+          (m/=> ~name ~schema)
+          defn#))))
 
+#?(:clj
+   (c/defn ->defn [schema report args]
+     (let [{:keys [name return doc meta arities] :as parsed} (m/parse schema args)
+           _            (when (= ::m/invalid parsed) (m/-fail! ::parse-error {:schema schema, :args args}))
+           parse        (fn [{:keys [args] :as parsed}] (merge (md/parse args) parsed))
+           ->schema     (fn [{:keys [schema]}] [:=> schema (:schema return :any)])
+           single       (= :single (key arities))
+           parglists    (if single (->> arities val parse vector) (->> arities val :arities (map parse)))
+           raw-arglists (map :raw-arglist parglists)
+           schema       (as-> (map ->schema parglists) $ (if single (first $) (into [:function] $)))]
+       `(def ~name (m/-instrument {:schema ~schema :report ~report}
+                     (c/fn ~name
+                       ~@(map (fn [{:keys [arglist prepost body]}] `(~arglist ~prepost ~@body)) parglists)
+                       ~@(when-not single (some->> arities val :meta vector))))))))
```

Comment:
I found a way to keep the meta with minimum effort, if you want to use it:
```suggestion
           schema       (as-> (map ->schema parglists) $ (if single (first $) (into [:function] $)))
           meta-to-defn (assoc meta :doc doc)]
       `(do
          (c/defn ~name ~meta-to-defn ~(map :arglist parglists))
          (set! ~name
            (m/-instrument {:schema ~schema :report ~report}
                           (c/fn ~name
                                 ~@(map (fn [{:keys [arglist prepost body]}] `(~arglist ~prepost ~@body)) parglists)
                                 ~@(when-not single (some->> arities val :meta vector)))))))))

```

But this code does not work with Clojure (because you can't `set!` root bindings). I'm supposing `->defn` is considered an implementation detail, so we can send an additional parameter (like `is-clojurescript?`) to handle the Clojure case?

Reply:
Thanks for looking into this! based on that I went with a strategy to produce an actual defn that just delegates to the instrumented version.